src/fsharp/TypeChecker.fs

```diff
@@ -725,13 +728,31 @@ let UnifyUnitType cenv denv m ty exprOpt =
         let resultTy = NewInferenceType ()
         if AddCxTypeEqualsTypeUndoIfFailed denv cenv.css m ty (domainTy --> resultTy) then 
             warning (FunctionValueUnexpected(denv,ty,m))
-        else
-            let perhapsProp = 
-                typeEquiv cenv.g cenv.g.bool_ty ty &&
+        else        
+            if not (typeEquiv cenv.g cenv.g.bool_ty ty) then 
+                warning (UnitTypeExpected (denv,ty,m)) 
+            else
                 match exprOpt with 
-                | Some(Expr.App(Expr.Val(vf,_,_),_,_,[__],_)) when vf.LogicalName = opNameEquals -> true
-                | _ -> false
-            warning (UnitTypeExpected (denv,ty,perhapsProp,m)) 
+                | Some(Expr.App(Expr.Val(vf,_,_),_,_,exprs,_)) when vf.LogicalName = opNameEquals ->
+                    match exprs with 
+                    | Expr.App(Expr.Val(prop,_,_),_,_,Expr.Val(vf,_,_) :: _,_) :: _ -> 
+                        let isProperty = prop.LogicalName.StartsWith "get_" // TODO: ask @dsyme about better way
+                        
+                        if isProperty then
+                            warning (UnitTypeExpectedWithPossiblePropertySetter (denv,ty,vf.LogicalName,m))
+                        else
+                            warning (UnitTypeExpectedWithEquality (denv,ty,m))
+                    | Expr.Op(_,_,Expr.Val(vf,_,_) :: _,_) :: _ -> 
+                        let isProperty = true // TODO: ask @dsyme about better way
```

Comment:
Looking in the MemberInfo of the F# value should do the trick, c.f. this in Symbols.fs.  You might want to add a new ``IsPropertyGetterMethod`` to ``Val`` and ``ValRef`` in ``TastOps.fs/fsi`` and make sure all similar chunks of code are shared.

```
        | V v -> 
            match v.MemberInfo with 
            | None -> false 
            | Some memInfo -> memInfo.MemberFlags.MemberKind = MemberKind.PropertyGet
```